### PR TITLE
ASoC: SOF: pcm: Improve debug and error prints

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

Our error printing alone leaves out information which can provide a needed context for the error itself.
For example in IPC3: "ipc tx error for 0x60010000 (msg/reply size: 108/20): -22"
This will not tell which PCM device, what direction this error had happened.

Similarly the debug prints are using ad-hoc formatting and they are not consistent, one needs to look for prior or later prints for context, which can be problematic with concurrent PCM operation.

At the heart this series will switch to a spcm_dbg() and spcm_err() wrapper to include useful information in prints where spcm is available.

As preparation, the PCM functions are relocated to pcm.c from sof-audio.c.

### CODE REVIEW

**Review by lgirdwood — DISMISSED** [review]
Nice one @ujfalusi - do you think similar can be done for kcontrols

**Comment by ujfalusi** [comment]
> Nice one @ujfalusi - do you think similar can be done for kcontrols

I'm going through the files and logs to see what can be improved. With a good log we can cut down on some back and forth to get the logs which have information and in case if it is a hard to reproduce issue, that can take a while...

File: sound/soc/sof/sof-audio.h [hunk 1]
```diff
@@ -617,6 +617,28 @@ struct snd_sof_pcm *snd_sof_find_spcm_comp(struct snd_soc_component *scomp,
 void snd_sof_pcm_period_elapsed(struct snd_pcm_substream *substream);
 void snd_sof_pcm_init_elapsed_work(struct work_struct *work);
 
+/*
+ * snd_sof_pcm specific wrappers for dev_dbg() and dev_err() to provide
+ * consistent and useful prints.
+ * The spcm_dbg() is prepared to handle a call without valid direction (dir == -1)
+ */
+#define spcm_dbg(__spcm, __dir, __fmt, ...) {					\
+	if (__dir == -1)							\
```
Comment (ranj063) [inline sound/soc/sof/sof-audio.h]: @ujfalusi this is really used only at the time of creation. Do we really need to special handle it just for that? In all practical cases, I think we're going to have the dir defined.

Comment (ujfalusi) [reply sound/soc/sof/sof-audio.h]: Yes, it is only in one case, I can put it as __unlikely(), printing the dir and a number would be confusing to look at if the direction is not valid?

Comment (ranj063) [reply sound/soc/sof/sof-audio.h]: no what i was suggesting was that we leave out the new macro for this one case and use it everywhere else with a valid direction.

**Comment by ujfalusi** [comment]
Changes since v1:
- use dev_dbg() directly at entry of sof_pcm_new and drop the __dir = -1 handling from the spcm_dbg() macro.

**Review by lgirdwood — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Comment by ranj063** [comment]
SOFCI TEST

**Comment by ranj063** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
> SOFCI TEST

The current results look as expected, I will rerun again though as some DUTs just got a BIOS update.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by ujfalusi** [comment]
SOFCI TEST

**Comment by ujfalusi** [comment]
@bardliao, @ranj063, @kv2019i, can we merge this?

**Comment by bardliao** [comment]
Test result looks good. Let's merge

**Pull request merged**